src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2884,6 +2884,206 @@ export default class EthereumApi implements Api {
+    // technically valid per the spec but we cannot go past the Genesis Block. Values
+    // above MAX_BLOCKS are technically within spec, however we cap totalBlocks because
+    // of the resource needs and potential abuse of a very large blockCount.
+    const totalBlocks = Math.min(
+      Math.max(
+        Quantity.toBigInt(blockCount) >= MAX_BLOCKS
+          ? MAX_BLOCKS
+          : Quantity.toNumber(blockCount),
+        MIN_BLOCKS
+      ),
+      Quantity.toNumber(
+        newestBlockNumber >= MAX_BLOCKS ? MAX_BLOCKS : newestBlockNumber + 1n
+      ),
+      MAX_BLOCKS
+    );
+
+    const baseFeePerGas: Quantity[] = new Array(totalBlocks);
+    const gasUsedRatio: number[] = new Array(totalBlocks);
+    let reward: Array<Quantity[]>;
+
+    // percentiles must be unique and in ascending order between 0 and 100
+    if (rewardPercentiles.length > 0) {
+      const ERR_INVALID_PERCENTILE =
+        "invalid reward percentile: percentiles must be unique and in ascending order";
+      if (rewardPercentiles[0] < 0)
+        throw new Error(`${ERR_INVALID_PERCENTILE} ${rewardPercentiles[0]}`);
+      if (rewardPercentiles[rewardPercentiles.length - 1] > 100)
+        throw new Error(
+          `${ERR_INVALID_PERCENTILE} ${
+            rewardPercentiles[rewardPercentiles.length - 1]
+          }`
+        );
+
+      for (let i = 1; i < rewardPercentiles.length; i++) {
+        if (rewardPercentiles[i] <= rewardPercentiles[i - 1]) {
+          throw new Error(
+            `${ERR_INVALID_PERCENTILE} ${rewardPercentiles[i]} ${
+              rewardPercentiles[i - 1]
+            }`
+          );
+        }
+      }
+
+      reward = new Array(totalBlocks);
+    }
+    // totalBlocks is inclusive of newestBlock
+    const oldestBlockNumber =
+      newestBlockNumber - Quantity.from(totalBlocks - 1).toBigInt();
+
+    let currentBlock: Block;
+    let currentPosition = 0;
+
+    while (currentPosition < totalBlocks) {
+      currentBlock = await blockchain.blocks.get(
+        Quantity.toBuffer(oldestBlockNumber + BigInt(currentPosition))
+      );
+
+      baseFeePerGas[currentPosition] = currentBlock.header.baseFeePerGas;
+
+      const gasUsed = currentBlock.header.gasUsed.toBigInt();
+      const gasLimit = currentBlock.header.gasLimit.toBigInt();
+
+      if (gasUsed === gasLimit) {
+        gasUsedRatio[currentPosition] = 1;
+      } else {
+        gasUsedRatio[currentPosition] = Number(
+          `0.${((gasUsed * PRECISION_BIG_INT) / gasLimit)
+            .toString()
+            .padStart(PAD_PRECISION, "0")}`
+        );
+      }
+
+      // For each percentile, find the cost of the unit of gas at that percentage
+      if (reward) {
+        const transactions = currentBlock.getTransactions();
+
+        // If there are no transactions, all reward percentiles are 0.
+        if (transactions.length === 0) {
+          reward[currentPosition] = rewardPercentiles.map(() => {
+            return Quantity.Zero;
+          });
```

Comment:
just a nit, but this is quite complex functionality, so I am in favour of removing anything that makes it harder for me to follow :)
```suggestion
          reward[currentPosition] = rewardPercentiles.map(() => Quantity.Zero);
```

Reply:
I updated a few other places where the `return` was similarly verbose